File: ZKCache/src/main/java/com/ericsson/ema/tim/dml/group/GroupBy.java
package com.ericsson.ema.tim.dml.group;

import com.ericsson.ema.tim.dml.DataTypes;
import com.ericsson.ema.tim.exception.DmlBadSyntaxException;
import com.ericsson.ema.tim.exception.DmlNoSuchFieldException;
import com.ericsson.ema.tim.dml.SelectClause;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Map;
import java.util.function.Function;

/**
 * Created by eqinson on 2017/4/23.
 */
public class GroupBy extends SelectClause {
    private final static Logger LOGGER = LoggerFactory.getLogger(GroupBy.class);

    private final String field;

    private GroupBy(String field) {
        this.field = field;
    }

    public static GroupBy groupBy(String field) {
        return new GroupBy(field);
    }

    public String getField() {
        return this.field;
    }

    public Function<Object, Object> grouping() {
        Map<String, String> metadata = getSelector().getContext().getTableMetadata();
        String fieldType = metadata.get(field);
        if (fieldType == null)
            throw new DmlNoSuchFieldException(field);

        switch (fieldType) {
            case DataTypes.String:
            case DataTypes.Int:
                return this::getFiledValFromTupleByName;
            default:
                LOGGER.error("unsupported data type: {},{}", field, fieldType);
                throw new DmlBadSyntaxException("Error: unsupported data type: " + field + "," + fieldType);
        }
    }
}

File: zkcache_scala/src/main/scala/com/ericsson/ema/tim/dml/group/GroupBy.scala
package com.ericsson.ema.tim.dml.group

import com.ericsson.ema.tim.dml.{DataTypes, SelectClause}
import com.ericsson.ema.tim.exception.{DmlBadSyntaxException, DmlNoSuchFieldException}
import org.slf4j.LoggerFactory

/**
  * Created by eqinson on 2017/5/13.
  */
class GroupBy(private val field: String) extends SelectClause {
	private val LOGGER = LoggerFactory.getLogger(classOf[GroupBy])

	override def getField: String = field

	type keyExtractorFuncType = (Object => Object)

	def keyExtractor(): keyExtractorFuncType = {
		selector.context.tableMetadata.get(field) match {
			case Some(DataTypes.String) | Some(DataTypes.Int) | Some(DataTypes.Boolean) =>
				o: Object => getFiledValFromTupleByName(o)
			case Some(otherType)                                                        =>
				LOGGER.error("unsupported data type: {},{}", field, otherType: Any)
				throw DmlBadSyntaxException("Error: unsupported data type: " + field + "," + otherType)
			case None                                                                   =>
				throw DmlNoSuchFieldException(field)

		}
	}
}


